# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/HTMLStyleEditor.cpp
# Commit: 6d9a0abd0a3c
# Full Hash: 6d9a0abd0a3c26f4337aff723c15795fe91fe884
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-27 09:19:35
# Description:
#   Bug 1870962 - Make `HTMLEditor::SplitAncestorStyledInlineElementsAt` never drop the "did split" state from the result r=m_kato
#   
#   Oddly, it returns the latest result as split result. However, in the case,
#   it splits `<kbd>` under `<button>`, but does not split `<button>`, but the
#   caller expects that whether it did split or not.
# ==============================================================================

diff -r 45dc475cd374 -r 6d9a0abd0a3c editor/libeditor/HTMLStyleEditor.cpp
--- a/editor/libeditor/HTMLStyleEditor.cpp	Tue Dec 26 22:56:56 2023 +0000
+++ b/editor/libeditor/HTMLStyleEditor.cpp	Wed Dec 27 02:53:44 2023 +0000
@@ -2147,6 +2147,7 @@
   AutoTArray<OwningNonNull<Element>, 24> arrayOfParents;
   for (Element* element :
        aPointToSplit.GetContainer()->InclusiveAncestorsOfType<Element>()) {
+    // XXX Cannot we stop if we meet a non-splittable element like <button>?
     if (HTMLEditUtils::IsBlockElement(
             *element, BlockInlineCheck::UseComputedDisplayOutsideStyle) ||
         !element->GetParent() ||
@@ -2285,8 +2286,10 @@
     if (!unwrappedSplitNodeResult.Handled()) {
       continue;
     }
-    // Mark the final result as handled forcibly.
-    result = unwrappedSplitNodeResult.ToHandledResult();
+    // Respect the last split result which actually did it.
+    if (!result.DidSplit() || unwrappedSplitNodeResult.DidSplit()) {
+      result = unwrappedSplitNodeResult.ToHandledResult();
+    }
     MOZ_ASSERT(result.Handled());
   }
 